### PR TITLE
SAK-48939 Samigo archive published assessments and question pools

### PR DESCRIPTION
**Pull request by smarquard**
Code originally from NYU and improved by UCT

### CODE REVIEW

File: samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/services/assessment/AssessmentEntityProducer.java [hunk 1]
```diff
@@ -406,4 +512,283 @@ public void updateEntityReferences(String toContext, Map<String, String> transve
...
+			} finally {
+				if (writer != null) {
+					try {
+						writer.close();
+					} catch (Throwable t) {
+					}
+				}
+			}
+		} catch (ParserConfigurationException | IdUnusedException e) {
+			e.printStackTrace();
```
Comment (jesusmmp) [inline samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/services/assessment/AssessmentEntityProducer.java:630]: log.info or log.error

File: samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/services/assessment/AssessmentEntityProducer.java [hunk 2]
```diff
@@ -406,4 +512,283 @@ public void updateEntityReferences(String toContext, Map<String, String> transve
...
+				// Pools used in published assessments (random draw)
+				loadResourceIds(db,
+					"select DISTINCT ENTRY FROM SAM_PUBLISHEDSECTIONMETADATA_T SM inner join SAM_PUBLISHEDSECTION_T SS on SM.SECTIONID = SS.SECTIONID " +
+					" where LABEL='POOLID_FOR_RANDOM_DRAW' and ASSESSMENTID = ?",
+					"ENTRY", assessmentId, result);
+			}
+
+		} catch (SQLException e) {
+			log.error(e.getMessage());
+			e.printStackTrace();
```
Comment (jesusmmp) [inline samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/services/assessment/AssessmentEntityProducer.java:787]: Remove

File: samigo/samigo-qti/src/java/org/sakaiproject/tool/assessment/qti/helper/assessment/AssessmentHelperBase.java [hunk 3]
```diff
@@ -456,17 +459,30 @@ public void updateIPAddressSet(Assessment assessmentXml,
...
-    	attachment.append("|");
-    	attachment.append(attachmentData.getFilename());
-    	attachment.append("|");
-    	attachment.append(attachmentData.getMimeType());
-    	attachment.append("\n");
+        Object elt = iter.next();
+
+        if (elt instanceof AttachmentData) {
+            AttachmentData attachmentData = (AttachmentData) elt;
+            attachment.append(attachmentData.getResourceId().replaceAll(" ", ""));
```
Comment (jesusmmp) [inline samigo/samigo-qti/src/java/org/sakaiproject/tool/assessment/qti/helper/assessment/AssessmentHelperBase.java:469]: Lines from 469 to 474 are the same 477 to 482
Better put them out of the if

File: samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/services/assessment/AssessmentEntityProducer.java [hunk 4]
```diff
@@ -406,4 +512,283 @@ public void updateEntityReferences(String toContext, Map<String, String> transve
...
+                            "select resourceid from SAM_PUBLISHEDATTACHMENT_T where itemtextid in " +
+                            " (select itemtextid from SAM_PUBLISHEDITEMTEXT_T where itemid in " +
+                            "  (select itemid from SAM_PUBLISHEDITEM_T where sectionid in" +
+                            "   (select sectionid from SAM_PUBLISHEDSECTION_T where assessmentid = ?)))",
+			    "resourceid",
+                            pubAssessmentId,
+                            result);
+        } catch (SQLException e) {
+            log.error(e.getMessage());
+            e.printStackTrace();
```
Comment (jesusmmp) [inline samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/services/assessment/AssessmentEntityProducer.java:726]: Remove

File: samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/services/assessment/AssessmentEntityProducer.java [hunk 5]
```diff
@@ -406,4 +512,283 @@ public void updateEntityReferences(String toContext, Map<String, String> transve
...
+				writer = new FileWriter(file);
+
+				TransformerFactory tFactory = TransformerFactory.newInstance();
+				Transformer transformer = tFactory.newTransformer();
+
+				DOMSource source = new DOMSource(doc);
+				StreamResult result = new StreamResult(writer);
+				transformer.transform(source, result);
+			} catch (IOException | TransformerException e) {
+				e.printStackTrace();
```
Comment (jesusmmp) [inline samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/services/assessment/AssessmentEntityProducer.java:620]: Remove

**Comment by smarquard** [comment]
Thanks for the review comments @jesusmmp - I'm aiming to pick this up again soon.

File: samigo/samigo-qti/src/java/org/sakaiproject/tool/assessment/qti/helper/assessment/AssessmentHelperBase.java (hunk 3 above)
Comment (smarquard) [reply samigo/samigo-qti/src/java/org/sakaiproject/tool/assessment/qti/helper/assessment/AssessmentHelperBase.java:469]: attachmentData is a different type in the two if cases.

Comment (smarquard) [reply samigo/samigo-qti/src/java/org/sakaiproject/tool/assessment/qti/helper/assessment/AssessmentHelperBase.java:469]: Seems like that would be more lines of code in total, so not sure it's worth it just to get rid of what looks likes fairly trivial duplication.

**Comment by smarquard** [comment]
@ern @ottenhoff  Thoughts on the direct SQL queries here? This is originally code we adapted from NYU so I'm not sure if it's there because using hibernate is not efficient given the data model or it was just a shortcut and there's a better way to do this.

**Comment by smarquard** [comment]
@ern  I've refactored this to get rid of the direct SQL queries. So it's ready for review.

File: samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/services/assessment/AssessmentEntityProducer.java [hunk 6]
```diff
@@ -27,13 +31,31 @@
...
 import org.apache.commons.lang3.StringUtils;
+import org.jsoup.Jsoup;
+import org.jsoup.select.Elements;
+
+import org.sakaiproject.authz.api.Member;
 import org.sakaiproject.component.api.ServerConfigurationService;
+import org.sakaiproject.component.cover.ComponentManager;
+import org.sakaiproject.content.api.ContentResource;
+import org.sakaiproject.content.cover.ContentHostingService;
+import org.sakaiproject.content.cover.ContentHostingService;
```
Comment (jesusmmp) [inline samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/services/assessment/AssessmentEntityProducer.java:57]: Duplicate line

File: samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/services/assessment/AssessmentEntityProducer.java [hunk 7]
```diff
@@ -53,15 +78,25 @@
...
+import org.sakaiproject.tool.assessment.data.dao.questionpool.QuestionPoolItemData;
 import org.sakaiproject.tool.assessment.facade.AssessmentFacade;
 import org.sakaiproject.tool.assessment.facade.PublishedAssessmentFacade;
 import org.sakaiproject.tool.assessment.facade.PublishedAssessmentFacadeQueries;
 import org.sakaiproject.tool.assessment.facade.PublishedAssessmentFacadeQueriesAPI;
 import org.sakaiproject.tool.assessment.facade.SectionFacade;
+import org.sakaiproject.tool.assessment.shared.api.questionpool.QuestionPoolServiceAPI;
 import org.sakaiproject.tool.assessment.shared.api.qti.QTIServiceAPI;
+import org.sakaiproject.user.api.UserNotDefinedException;
+import org.sakaiproject.user.cover.UserDirectoryService;
```
Comment (jesusmmp) [inline samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/services/assessment/AssessmentEntityProducer.java:92]: We should use import org.sakaiproject.user.api.UserDirectoryService; if is possible

File: samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/services/assessment/AssessmentEntityProducer.java [hunk 8]
```diff
@@ -393,4 +520,284 @@ public void updateEntityReferences(String toContext, Map<String, String> transve
...
+				result.add(linkRef);
+				log.info("Found inline public asset: {} adding to attachment list as: {}", link, linkRef);
+			}
+
+		} catch (UnsupportedEncodingException e) {
+			// Don't expect this ever to happen
+			log.warn("Unable to add link {} to attachment list: {}", link, e.getMessage());
+		} catch (UserNotDefinedException e) {
+			log.warn("Unable to add link {} to attachment list", link);
+			log.error("Trace", e);
```
Comment (jesusmmp) [inline samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/services/assessment/AssessmentEntityProducer.java:755]: Remove

File: samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/services/assessment/AssessmentEntityProducer.java [hunk 9]
```diff
@@ -393,4 +520,284 @@ public void updateEntityReferences(String toContext, Map<String, String> transve
...
+			}
+
+			// references to public files
+			if (link.contains("/access/content/public/")) {
+				String linkRef = URLDecoder.decode(link.replace(link.substring(0, link.indexOf("/public/")), ""), "UTF-8");
+				result.add(linkRef);
+				log.info("Found inline public asset: {} adding to attachment list as: {}", link, linkRef);
+			}
+
+		} catch (UnsupportedEncodingException e) {
```
Comment (jesusmmp) [inline samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/services/assessment/AssessmentEntityProducer.java:750]: Only one catch with both and the same message?

log.warn("Unable to add link {} to attachment list: {}", link, e.getMessage());

File: samigo/samigo-qti/src/java/org/sakaiproject/tool/assessment/qti/helper/AuthoringHelper.java [hunk 10]
```diff
@@ -315,6 +327,146 @@ public Document getAssessment(String assessmentId, InputStream is)
...
+      Set securedIPAddressSet = (Set) publishedAssessment.getSecuredIPAddressSet();
+      if (securedIPAddressSet != null)
+      {
+        assessmentHelper.updateIPAddressSet(assessmentXml,
+                securedIPAddressSet);
+      }
+
+      Set attachmentSet = (Set) publishedAssessment.getAssessmentAttachmentSet();
+
+      if (attachmentSet != null && attachmentSet.size() != 0)
```
Comment (jesusmmp) [inline samigo/samigo-qti/src/java/org/sakaiproject/tool/assessment/qti/helper/AuthoringHelper.java:439]: if (!attachmentSet.isEmpty()) {

File: samigo/samigo-qti/src/java/org/sakaiproject/tool/assessment/qti/helper/assessment/AssessmentHelperBase.java [hunk 11]
```diff
@@ -557,6 +630,24 @@ private void setField(Assessment assessmentXml, AssessmentFacade assessment,
...
+  private void setField(Assessment assessmentXml, PublishedAssessmentFacade assessment,
+                        String key, String translationKey)
+  {
+
+    String value = assessment.getAssessmentMetaDataByLabel(key);
+    log.debug("setField(Assessment assessmentXml, AssessmentFacade assessment, String key, String translationKey)");
+    log.debug("key: " + key);
+    log.debug("value: " + value);
+
+    if (value == null)
```
Comment (jesusmmp) [inline samigo/samigo-qti/src/java/org/sakaiproject/tool/assessment/qti/helper/assessment/AssessmentHelperBase.java:642]: assessmentXml.setFieldentry(translationKey, value != null ? value : "false");

File: samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/services/assessment/AssessmentEntityProducer.java (hunk 6 above)
Comment (jesusmmp) [reply samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/services/assessment/AssessmentEntityProducer.java:57]: Also, we should use import org.sakaiproject.content.api.ContentHostingService; if is possible

**Comment by smarquard** [comment]
Anything further to do here, @jesusmmp and @ern  ?

**Review by jesusmmp — APPROVED** [review]
LGTM!

**Pull request merged**